test/ExpressiveInstallerTest/RoutersTest.php

```diff
@@ -93,9 +106,27 @@ public function testRouter(
         );
 
         // Test home page
-        $setupRoutes = (strpos($copyFilesKey, 'minimal') !== 0);
+        $setupRoutes = strpos($copyFilesKey, 'minimal') !== 0;
         $response = $this->getAppResponse('/', $setupRoutes);
         $this->assertEquals($expectedResponseStatusCode, $response->getStatusCode());
+
+        /** @var Application $app */
+        $app = $container->get(Application::class);
+        $this->assertCount(count($expectedRoutes), $app->getRoutes());
+        foreach ($app->getRoutes() as $route) {
+            foreach ($expectedRoutes as $expectedRoute) {
+                if ($expectedRoute['name'] === $route->getName()) {
+                    $this->assertEquals($expectedRoute['path'], $route->getPath());
+                    // todo: not sure how we can test middleware?
+                    // $this->assertEquals($expectedRoute['middleware'], $route->getMiddleware());
```

Comment:
> How do we know then if correct middleware is executed?

You check the response code so you know it executed ok. And you can check the response body for specific strings to make sure the correct action is executed. I think it doesn't matter what exactly is executed as long as you get the right code and content. To make sure the right content is generated in an action you need to test each action separately.

With those checks, I think you can be sure it works as intended.